src/rpc/rewards.cpp

```diff
@@ -0,0 +1,883 @@
+            {   "rewards",      "getpayments",                &getpayments,                {"reward_id"}},
+            {   "rewards",      "cancelpayments",             &cancelpayments,             {"reward_id"}},
+            {   "rewards",      "executepayments",            &executepayments,            {"reward_id"}},
+    };
+
+void RegisterRewardsRPCCommands(CRPCTable &t)
+{
+    for (unsigned int vcidx = 0; vcidx < ARRAYLEN(commands); vcidx++)
+        t.appendCommand(commands[vcidx].name, &commands[vcidx]);
+}
+
+CAmount AmountFromValue(bool p_isRVN, const UniValue& p_value)
+{
+    if (!p_value.isNum() && !p_value.isStr())
+        throw JSONRPCError(RPC_TYPE_ERROR, "Amount is not a number or string");
+
+    CAmount amount;
+    if (!ParseFixedPoint(p_value.getValStr(), 8, &amount))
+        throw JSONRPCError(RPC_TYPE_ERROR, strprintf("Invalid amount (3): %s", p_value.getValStr()));
+
+    if (p_isRVN && !MoneyRange(amount))
+        throw JSONRPCError(RPC_TYPE_ERROR, strprintf("Amount out of range: %s", amount));
+
+    return amount;
+}
+
+bool GenerateTransaction(
+    CWallet * const p_walletPtr, const std::string & p_src,
+    std::vector<CPayment> & p_payments,
+    std::vector<PendingTransaction> & p_pendingTxns)
+{
+    bool fcnRetVal = false;
+    size_t expectedCount = 0;
+    size_t actualCount = 0;
+
+    LogPrintf("Generating transactions for payments...\n");
+
+    // While condition is false to ensure a single pass through this logic
+    do {
+        //  Transfer the specified amount of the asset from the source to the target
+        CCoinControl ctrl;
+        std::shared_ptr<CWalletTx> txnPtr = std::make_shared<CWalletTx>();
+        std::shared_ptr<CReserveKey> reserveKeyPtr = std::make_shared<CReserveKey>(p_walletPtr);
+        std::shared_ptr<UniValue> batchResult = std::make_shared<UniValue>(UniValue::VOBJ);
+        CAmount nFeeRequired = 0;
+        CAmount totalPaymentAmt = 0;
+
+        //  Handle payouts using RVN differently from those using an asset
+        if (p_src == "RVN") {
+            // Check amount
+            CAmount curBalance = p_walletPtr->GetBalance();
+
+            if (p_walletPtr->GetBroadcastTransactions() && !g_connman) {
+                LogPrintf("Error: Peer-to-peer functionality missing or disabled\n");
+                break;
+            }
+
+            std::vector<CRecipient> vDestinations;
+
+            //  This should (due to external logic) only include pending payments
+            for (auto & payment : p_payments) {
+                expectedCount++;
+
+                // Parse Raven address
+                CTxDestination dest = DecodeDestination(payment.address);
+                if (!IsValidDestination(dest)) {
+                    LogPrintf("Destination address '%s' is invalid.\n", payment.address.c_str());
+                    payment.completed = true;
+                    continue;
+                }
+
```

Comment:
Make sure all address are valid when creating the snapshot.

Reply:
Ownership addresses are now validated during snapshot creation.